crates/lang_srv/src/server.rs

```diff
@@ -1,34 +1,54 @@
 use analysis::HIGHLIGHT_TOKENS_LEGEND;
-use parking_lot::{Mutex, MutexGuard};
-use registry::{DocumentChange, Registry};
+use registry::Registry;
+use std::future::Future;
+
+use std::sync::Arc;
+use std::time::Duration;
 use tower_lsp::jsonrpc::Result;
 use tower_lsp::lsp_types::*;
 use tower_lsp::{Client, LanguageServer, LspService, Server};
 
+use crate::analysis::global_analysis;
+
 mod analysis;
 mod convert;
 mod registry;
 
 #[derive(Debug)]
 struct RocLs {
+    pub inner: Arc<Inner>,
+}
+#[derive(Debug)]
+struct Inner {
     client: Client,
-    registry: Mutex<Registry>,
+    registry: Registry,
+    change_cancel_handle: parking_lot::Mutex<(
+        tokio::sync::watch::Sender<i32>,
+        tokio::sync::watch::Receiver<i32>,
+    )>,
+    documents_updating: tokio::sync::Semaphore,
 }
 
 impl std::panic::RefUnwindSafe for RocLs {}
+const SEMLIMIT: u32 = 20;
 
 impl RocLs {
     pub fn new(client: Client) -> Self {
         Self {
-            client,
-            registry: Mutex::new(Registry::default()),
+            inner: Arc::new(Inner {
+                client,
+                registry: Registry::default(),
+                change_cancel_handle: parking_lot::Mutex::new(tokio::sync::watch::channel(0)),
+                ///Used for identifying if the intial stage of a document update is complete
+                documents_updating: tokio::sync::Semaphore::new(SEMLIMIT as usize),
+            }),
         }
     }
-
-    fn registry(&self) -> MutexGuard<Registry> {
-        self.registry.lock()
+    ///Wait for all the semaphores associated with an in-progress document_info update to be released
+    async fn wait_for_changes(&self) {
+        //We don't actually need to return the permit, all we want is to have it momenterily which guarentees that any updates started before the call to this are complete
+        drop(self.inner.documents_updating.acquire_many(SEMLIMIT).await.expect("semaphore permits could not be aquired for the document updating process. This is unrecoverable"));
     }
```

Comment:
I'd really like us to remove this. I don't think we should ever be blocking in this way if the language server will support lazy background updates; I think it's best to avoid concurrent updates right now, and when needed, documents should be able to update independently.